### PR TITLE
CheckIt Tests Part II

### PR DESCRIPTION
**Pull request by andrayantelo**
Wrote a few tests but for the most part ended up refactoring checkit methods.

### CODE REVIEW

File: docs/checkit.js [hunk 1]
```diff
@@ -438,22 +441,175 @@ CheckIt.prototype.removeEmptyWeeks = function(calObj, $calendarDiv) {
...
-    this.$fullForm[0].reset();
+};
+
+CheckIt.prototype.removeFieldError = function($id, $srId) {
+    // Removes error state CSS on form field 
+    // $id is the selector for the div with class .form-group that
+    // surrounds the input field with the error
+    // $srId is the selector for the screen reader element
+    $id.removeClass('has-error has-feedback');
+    $srId.addClass('hidden');
```
Comment (johntyree) [inline docs/checkit.js:486]: No aria here?

File: docs/checkit.js [hunk 2]
```diff
@@ -438,22 +441,175 @@ CheckIt.prototype.removeEmptyWeeks = function(calObj, $calendarDiv) {
...
+};
+
+CheckIt.prototype.addHelpBlock = function($helpBlock) {
```
Comment (johntyree) [inline docs/checkit.js:489]: Looks like these four add/remove functions can be combined to just two show and hide functions.

File: docs/checkit.js [hunk 3]
```diff
@@ -438,22 +441,175 @@ CheckIt.prototype.removeEmptyWeeks = function(calObj, $calendarDiv) {
...
+    var $helpBlock = $formGroup.find('#helpBlock');
+    var $fiveYears = $formGroup.find('#fiveYears');
+    var difference = endDate.diff(startDate.format("YYYY-MM-DD"), 'years', true);
+    
+    if (startDate.isBefore(endDate)) {
+        // startDate is before endDate so errors can be removed.
+        this.removeFieldError($formGroup, $sr);
+        this.removeHelpBlock($helpBlock);
+        // If there are more than 5 years between the dates return false for invalid
+        if (difference < 5) {
```
Comment (johntyree) [inline docs/checkit.js:552]: Why have this?

File: docs/checkit.js [hunk 4]
```diff
@@ -667,17 +725,17 @@ CheckIt.prototype.buildCalendar = function(calendarObject) {
...
-    this.clearPage();
+    this.clearPage(this.$calendarDiv);
     this.buildCalendar(calendarObj);
     this.store.setActiveById(calendarObj.state.uniqueId);
     this.store.save(calendarObj);
 };
 
-CheckIt.prototype.clearPage = function() {
+CheckIt.prototype.clearPage = function($div) {
     // Remove all divs from page except #template
```
Comment (johntyree) [inline docs/checkit.js]: This comment is not accurate.

File: tests/tests.js [hunk 5]
```diff
@@ -439,79 +439,275 @@ QUnit.test("hideLoadingWheel test", function( assert ) {
...
 });
 
-QUnit.test("clearForm test", function( assert ) {
-    assert.expect(0);
-});
 
 QUnit.test("addCalendarToDropdown test", function( assert ) {
-    assert.expect(0);
+    assert.expect(2);
+    this.checkit.addCalendarToDropdown("1234", "Hello", this.$dropdown);
```
Comment (johntyree) [inline tests/tests.js]: does this still work if there are other entries in the dropdown already?

File: tests/tests.js [hunk 6]
```diff
@@ -439,79 +439,275 @@ QUnit.test("hideLoadingWheel test", function( assert ) {
...
+    assert.expect(1);
+    
+    this.$calendarDiv.append(`<div id="testCollapse" class="collapse"></div>`);
+    
+    var $testCollapse = this.$calendarDiv.find('#testCollapse');
+    assert.equal($testCollapse.css('display'), 'none');
+    this.checkit.showForm($testCollapse);
+    setTimeout(function(){ 
+        assert.equal($testCollapse.css('display'), 'block');
+        }, 350);
```
Comment (johntyree) [inline tests/tests.js:540]: This works kind of, but it's basically a race condition. What happened to using the transitionend event?

**Review by johntyree — APPROVED** [review]
It was pretty hard to slog through reading these tests.

**Review by johntyree — CHANGES_REQUESTED** [review]
Accidental approval.

File: docs/checkit.js (hunk 1 above)
Comment (andrayantelo) [reply docs/checkit.js:486]: What do you mean? srId is the selector for the screen reader element.

File: docs/checkit.js (hunk 3 above)
Comment (andrayantelo) [reply docs/checkit.js:552]: so that people don't make a calendar that is 500 years long. Or a million.

File: docs/checkit.js (hunk 4 above)
Comment (andrayantelo) [reply docs/checkit.js]: Removed.

File: tests/tests.js (hunk 6 above)
Comment (andrayantelo) [reply tests/tests.js:540]: It wouldn't 'hear' the transitionend event, which is why i switched to using setTimeout.

File: docs/checkit.js (hunk 2 above)
Comment (andrayantelo) [reply docs/checkit.js:489]: Didn't do this because of reasons we talked about.

File: tests/tests.js (hunk 5 above)
Comment (andrayantelo) [reply tests/tests.js]: Added testing for that case.

File: tests/tests.js (hunk 6 above)
Comment (johntyree) [reply tests/tests.js:540]: That's odd. You were attaching the event listener to `$testCollapse` I assume?

File: docs/checkit.js [hunk 7]
```diff
@@ -732,7 +732,7 @@ CheckIt.prototype.displayCalendar = function(calendarObj) {
 };
 
 CheckIt.prototype.clearPage = function($div) {
-    // Remove all divs from page except #template
+    // Remove all calendar related divs from page
```
Comment (johntyree) [inline docs/checkit.js]: this is still wrong. It removes the checkit related divs from inside `$div`. The rest are untouched.

File: tests/tests.js [hunk 8]
```diff
@@ -463,10 +463,15 @@ QUnit.module( "CheckIt tests dropdown", {
 
 
 QUnit.test("addCalendarToDropdown test", function( assert ) {
-    assert.expect(2);
+    assert.expect(3);
     this.checkit.addCalendarToDropdown("1234", "Hello", this.$dropdown);
     assert.equal(this.$dropdown.children().attr('id'), '1234');
     assert.equal(this.$dropdown.children().text(), "Hello");
+    
+    // Check with items in dropdown already.
```
Comment (johntyree) [inline tests/tests.js]: I meant the previous asserts with something in the dropdown already. I suggest.

1. Add made up thing to dropdown
2. Check that 'hello' whatever isn't there.
3. Add "1234", "Hello" to dropdown.
4. Test that it *is* there.

**Review by johntyree — CHANGES_REQUESTED** [review]

File: docs/checkit.js (hunk 7 above)
Comment (andrayantelo) [reply docs/checkit.js]: Fixed.

File: tests/tests.js (hunk 8 above)
Comment (andrayantelo) [reply tests/tests.js]: Fixed.

File: tests/tests.js [hunk 9]
```diff
@@ -472,6 +472,12 @@ QUnit.test("addCalendarToDropdown test", function( assert ) {
     this.$dropdown.append("<li>Hello World</li>");
     this.checkit.addCalendarToDropdown("101", "Cal", this.$dropdown);
     assert.equal(this.$dropdown.children().length, 3);
+    
+    this.$dropdown.empty()
```
Comment (johntyree) [inline tests/tests.js]: Why do you clear it? Now I don't understand what this is supposed to be doing.

**Review by johntyree — APPROVED** [review]

**Review by johntyree — CHANGES_REQUESTED** [review]
Accidentally approved again. Github really needs to make this less annoying to use correctly.

File: tests/tests.js (hunk 9 above)
Comment (andrayantelo) [reply tests/tests.js]: Fixed.

File: tests/tests.js (hunk 6 above)
Comment (andrayantelo) [reply tests/tests.js:540]: If I remember correctly, yes. This was at the beginning of this week.

**Review by johntyree — APPROVED** [review]

**Pull request merged**